packages/ui/src/components/Visualization/Visualization.tsx

```diff
@@ -14,15 +16,18 @@ interface CanvasProps {
 }
 
 export const Visualization: FunctionComponent<PropsWithChildren<CanvasProps>> = (props) => {
+  const controller = useMemo(() => ControllerService.createController(), []);
   const lastUpdate = useMemo(() => Date.now(), [props.entities]);
 
   return (
-    <div className={`canvas-surface ${props.className ?? ''}`}>
-      <ErrorBoundary key={lastUpdate} fallback={props.fallback ?? <CanvasFallback />}>
+    <VisualizationProvider controller={controller}>
```

Comment:
Moving the `VisualizationProvider` here, to avoid recreating it whenever the `entities` array change.

cc @jeff-phillips-18